---
title: Analysis of smoking status impact on molecular mechanisms of SARS -CoV-2 viral
  entry through single-cell sequencing experiments.
author: "Mateusz Urbanek"
date: "12/19/2021"
output: pdf_document
---

```{r setup, message=FALSE, warning=FALSE}
# package used for plotting
if (!require("ggplot2")) { install.packages("ggplot2") }

# package used for loading big data sets
if (!require("data.table")) { install.packages("data.table") }

# requires libcurl_dev
if (!require("tidyverse")) { install.packages("tidyverse") }

if (!require("geometry")) { install.packages("geometry") }

if (!require("foreach")) { install.packages("foreach") }
if (!require("parallel")) { install.packages("parallel") }
if (!require("doParallel")) { install.packages("doParallel") }

if (!require("umap")) { install.packages("umap") }

if (!require("Seurat")) { install.packages("Seurat") }
if (!require("hdf5r")) { install.packages("hdf5r") }

if (!require("BiocManager")) { 
  install.packages("BiocManager")
  BiocManager::install()  
}

BiocManager::install(c("SingleCellExperiment", "scater", "GEOquery", "rhdf5", "DropletUtils"))
```

```{r read ExpressionSet file, message=TRUE, warning=TRUE}
library("GEOquery")

gse = getGEO(filename="Data/GSE/GSE122960_series_matrix.txt", GSEMatrix=TRUE, getGPL = FALSE)

x <- phenoData(gse)
pData(x)
```

```{r split path function, message=FALSE, warning=FALSE}
split_path <- function(x) if (dirname(x)==x) x else c(basename(x),split_path(dirname(x)))
```

```{r read data, message=FALSE, warning=FALSE}
path.data.raw <- "./Data/GSE/RAW/"

for (i in 1:length(gse$title)) {
  a = split_path(gse$supplementary_file_1[i])
  b = split_path(gse$supplementary_file_2[i])
  print(a[1])
  
  f1 = paste(path.data.raw, a[1], sep='')
  if (!file.exists(f1)) {
    download.file(gse$supplementary_file_1[i], f1)
  }
  
  print(b[1])
  f2 = paste(path.data.raw, b[1], sep='')
  if (!file.exists(f2)) {
    download.file(gse$supplementary_file_2[i], f2)
  }
  
  print("====================")
}
```

```{r read data, message=TRUE, warning=TRUE}
library(dplyr)
library(Seurat)
library(patchwork)

# sce = c()
# for (f in list.files(path.data.raw, pattern = "*.h5")) {
#   if (grepl("raw", f, fixed=TRUE)) {
#     print(f)
#     sce <- append(sce, Read10X_h5(filename = paste(path.data.raw, f, sep = "")))
#   }
# }

for (f in list.files(path.data.raw, pattern = "*.h5")) {
  if (grepl("raw", f, fixed=TRUE)) {
    pbmc.data <- Read10X_h5(filename = paste(path.data.raw, f, sep = ""))
    break
  }
}

pbmc <- CreateSeuratObject(counts = pbmc.data)
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")
pbmc[["percent.mt"]]

VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", scale.factor = 10000)

pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(pbmc), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(pbmc)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2

all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes)

pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))

print(pbmc[["pca"]], dims = 1:5, nfeatures = 5)

VizDimLoadings(pbmc, dims = 1:2, reduction = "pca")

DimPlot(pbmc, reduction = "pca")

DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)

DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)

pbmc <- JackStraw(pbmc, num.replicate = 100)
pbmc <- ScoreJackStraw(pbmc, dims = 1:20)

JackStrawPlot(pbmc, dims = 1:15)

pbmc <- FindNeighbors(pbmc, dims = 1:10)
pbmc <- FindClusters(pbmc, resolution = 0.5)

head(Idents(pbmc), 5)

pbmc <- RunUMAP(pbmc, dims = 1:10)

DimPlot(pbmc, reduction = "umap")

```

