{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6dd19221-da4b-43a5-a720-90a082ff3e14",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.10 (BiocManager 1.30.16), R 3.6.3 (2020-02-29)\n",
      "\n",
      "Installing package(s) 'BiocVersion'\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Old packages: 'backports', 'blob', 'brio', 'bslib', 'cachem', 'caret', 'cli',\n",
      "  'colorspace', 'crayon', 'curl', 'data.table', 'DBI', 'desc', 'devtools',\n",
      "  'diffobj', 'digest', 'dplyr', 'e1071', 'fansi', 'forecast', 'fs', 'generics',\n",
      "  'ggplot2', 'git2r', 'glue', 'htmltools', 'htmlwidgets', 'httpuv', 'ipred',\n",
      "  'isoband', 'knitr', 'later', 'lattice', 'lava', 'lifecycle', 'lmtest',\n",
      "  'lubridate', 'Matrix', 'memoise', 'mgcv', 'mime', 'nlme', 'openssl',\n",
      "  'pbdZMQ', 'pillar', 'pkgbuild', 'pkgload', 'pROC', 'proxy', 'R6',\n",
      "  'rcmdcheck', 'Rcpp', 'RcppArmadillo', 'RCurl', 'recipes', 'remotes', 'rlang',\n",
      "  'roxygen2', 'RSQLite', 'sessioninfo', 'shiny', 'stringi', 'survival',\n",
      "  'testthat', 'tibble', 'tidyr', 'tseries', 'TTR', 'usethis', 'utf8', 'uuid',\n",
      "  'waldo', 'withr', 'xfun', 'xml2'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if (!require(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\")\n",
    "BiocManager::install()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cf000c8a-093e-47dc-b85d-b8f7fa7386cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.10 (BiocManager 1.30.16), R 3.6.3 (2020-02-29)\n",
      "\n",
      "Installing package(s) 'GEOquery'\n",
      "\n",
      "also installing the dependencies ‘progress’, ‘BiocGenerics’, ‘hms’, ‘vroom’, ‘cpp11’, ‘tzdb’, ‘Biobase’, ‘readr’, ‘limma’\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Old packages: 'backports', 'blob', 'brio', 'bslib', 'cachem', 'caret', 'cli',\n",
      "  'colorspace', 'crayon', 'curl', 'data.table', 'DBI', 'desc', 'devtools',\n",
      "  'diffobj', 'digest', 'dplyr', 'e1071', 'fansi', 'forecast', 'fs', 'generics',\n",
      "  'ggplot2', 'git2r', 'glue', 'htmltools', 'htmlwidgets', 'httpuv', 'ipred',\n",
      "  'isoband', 'knitr', 'later', 'lattice', 'lava', 'lifecycle', 'lmtest',\n",
      "  'lubridate', 'Matrix', 'memoise', 'mgcv', 'mime', 'nlme', 'openssl',\n",
      "  'pbdZMQ', 'pillar', 'pkgbuild', 'pkgload', 'pROC', 'proxy', 'R6',\n",
      "  'rcmdcheck', 'Rcpp', 'RcppArmadillo', 'RCurl', 'recipes', 'remotes', 'rlang',\n",
      "  'roxygen2', 'RSQLite', 'sessioninfo', 'shiny', 'stringi', 'survival',\n",
      "  'testthat', 'tibble', 'tidyr', 'tseries', 'TTR', 'usethis', 'utf8', 'uuid',\n",
      "  'waldo', 'withr', 'xfun', 'xml2'\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Setting options('download.file.method.GEOquery'='auto')\n",
      "\n",
      "Setting options('GEOquery.inmemory.gpl'=FALSE)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "BiocManager::install(c(\"GEOquery\"))\n",
    "library(\"GEOquery\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "19fab433-0368-416a-9135-2efd03715116",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[1mRows: \u001b[1m\u001b[22m\u001b[34m\u001b[34m0\u001b[34m\u001b[39m \u001b[1m\u001b[1mColumns: \u001b[1m\u001b[22m\u001b[34m\u001b[34m18\u001b[34m\u001b[39m\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1m\u001b[1mColumn specification\u001b[1m\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (18): ID_REF, GSM3489182, GSM3489183, GSM3489184, GSM3489185, GSM3489186...\n",
      "\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use \u001b[30m\u001b[47m\u001b[30m\u001b[47m`spec()`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set \u001b[30m\u001b[47m\u001b[30m\u001b[47m`show_col_types = FALSE`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to quiet this message.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "gse = getGEO(filename=\"Data/GSE/GSE122960_series_matrix.txt\", GSEMatrix=TRUE, getGPL = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "467d73e6-0adc-4fd5-b7df-bff3aaaa0a70",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ExpressionSet (storageMode: lockedEnvironment)\n",
       "assayData: 0 features, 17 samples \n",
       "  element names: exprs \n",
       "protocolData: none\n",
       "phenoData\n",
       "  sampleNames: GSM3489182 GSM3489183 ... GSM3489198 (17 total)\n",
       "  varLabels: title geo_accession ... disease condition:ch1 (43 total)\n",
       "  varMetadata: labelDescription\n",
       "featureData: none\n",
       "experimentData: use 'experimentData(object)'\n",
       "  pubMedIds: 30554520 \n",
       "Annotation: GPL20301 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "813c38dd-d090-4728-84fd-cfea9f50405e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Experiment data\n",
       "  Experimenter name: Alexander,V,Misharin \n",
       "  Laboratory:  \n",
       "  Contact information: a-misharin@northwestern.edu \n",
       "  Title: Single-Cell Transcriptomic Analysis of Human Lung Reveals Complex Multicellular Changes During Pulmonary Fibrosis II \n",
       "  URL: https://www.atsjournals.org/doi/abs/10.1164/rccm.201712-2410OC \n",
       "  PMIDs: 30554520 \n",
       "\n",
       "  Abstract: A 124 word abstract is available. Use 'abstract' method.\n",
       "  notes:\n",
       "   citation:     \n",
       "      Paul A Reyfman, James M Walter, Nikita Joshi, Kishore R Anekalla, Alexandr\n",
       "a C McQuattie-Pimentel, Stephen Chiu, Ramiro Fernandez, Mahzad Akbarpour, \n",
       "Ching-I Chen, Ziyou Ren, Rohan Verma, Hiam Abdala-Valencia, Kiwon Nam, Mon\n",
       "ica Chi, SeungHye Han, Francisco J Gonzalez-Gonzalez, Saul Soberanes, Sato\n",
       "shi Watanabe, Kinola J.N. Williams, Annette S Flozak, Trevor T Nicholson, \n",
       "Vince K Morgan, Deborah R Winter, Monique Hinchcliff, Cara L Hrusch, Rober\n",
       "t D Guzy, Catherine A Bonham, Anne I Sperling, Remzi Bag, Robert B Hamanak\n",
       "a, Gökhan M Mutlu, Anjana V Yeldandi, Stacy A Marshall, Ali Shilatifard, L\n",
       "uis A.N. Amaral, Harris Perlman, Jacob I. Sznajder, A. Christine Argento, \n",
       "Colin T Gillespie, Jane Dematte, Manu Jain, Benjamin D Singer, Karen M Rid\n",
       "ge, Anna P Lam, Ankit Bharat, Sangeeta M. Bhorade, Cara J. Gottardi, G.R. \n",
       "Scott Budinger, and Alexander V Misharin. Single-Cell Transcriptomic Analy\n",
       "sis of Human Lung Provides Insights into the Pathobiology of Pulmonary Fib\n",
       "rosis. Am J Respir Crit Care Med. 2018. doi:10.1164/rccm.201712-2410OC\n",
       "   contact_address:     \n",
       "      240 E Huron\n",
       "   contact_city:     \n",
       "      Chicago\n",
       "   contact_country:     \n",
       "      USA\n",
       "   contact_department:     \n",
       "      Pulmonary and Critical Care\n",
       "   contact_email:     \n",
       "      a-misharin@northwestern.edu\n",
       "   contact_institute:     \n",
       "      Northwestern University\n",
       "   contact_laboratory:     \n",
       "      Misharin lab\n",
       "   contact_name:     \n",
       "      Alexander,V,Misharin\n",
       "   contact_phone:     \n",
       "      3125031958\n",
       "   contact_state:     \n",
       "      IL\n",
       "   contact_zip/postal_code:     \n",
       "      60611\n",
       "   contributor:     \n",
       "      Paul,A,Reyfman\n",
       "James,M,Walter\n",
       "Nikita,,Joshi\n",
       "Alexandra,C,McQuattie\n",
       "Kishore,\n",
       ",Anekalla\n",
       "GR,S,Budinger\n",
       "Alexander,V,Misharin\n",
       "   geo_accession:     \n",
       "      GSE122960\n",
       "   last_update_date:     \n",
       "      Mar 21 2019\n",
       "   overall_design:     \n",
       "      Single-cell suspensions from eight donor lungs and nine lungs from patient\n",
       "s with various forms of pulmonary fibrosis were prepared and single cell R\n",
       "NA-seq libraries were generated using 3' V2 chemistry kit on Chromium Sing\n",
       "le cell controller (10x Genomics). This GEO submission contains only proce\n",
       "ssed data (raw counts tables in HDF5 format), raw data (FASTQ files) avail\n",
       "able from dbGaP/SRA (phs001750.v1.p1).\n",
       "   platform_id:     \n",
       "      GPL20301\n",
       "   platform_taxid:     \n",
       "      9606\n",
       "   pubmed_id:     \n",
       "      30554520\n",
       "   relation:     \n",
       "      BioProject: https://www.ncbi.nlm.nih.gov/bioproject/PRJNA507000\n",
       "   sample_id:     \n",
       "      GSM3489182 GSM3489183 GSM3489184 GSM3489185 GSM3489186 GSM3489187 GSM34891\n",
       "88 GSM3489189 GSM3489190 GSM3489191 GSM3489192 GSM3489193 GSM3489194 GSM34\n",
       "89195 GSM3489196 GSM3489197 GSM3489198 \n",
       "   sample_taxid:     \n",
       "      9606\n",
       "   status:     \n",
       "      Public on Dec 15 2018\n",
       "   submission_date:     \n",
       "      Nov 26 2018\n",
       "   summary:     \n",
       "      Pulmonary fibrosis is a heterogeneous syndrome in which fibrotic scar repl\n",
       "aces normal lung tissue. We performed massively parallel single-cell RNA-S\n",
       "eq on lung tissue from eight lung transplant donors and eight patients wit\n",
       "h pulmonary fibrosis. Combined with in situ RNA hybridization, with amplif\n",
       "ication, these data provide a molecular atlas of disease pathobiology. We \n",
       "identified a distinct, novel population of profibrotic alveolar macrophage\n",
       "s exclusively in patients with fibrosis. Within epithelial cells, the expr\n",
       "ession of genes involved in Wnt secretion and response was restricted to n\n",
       "on-overlapping cells. We identified rare cell populations including airway\n",
       " stem cells and senescent cells emerging during pulmonary fibrosis. Analys\n",
       "is of a cryobiopsy specimen from a patient with early disease supports the\n",
       " clinical application of single-cell RNA-Seq to develop personalized appro\n",
       "aches to therapy.\n",
       "   supplementary_file:     \n",
       "      ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE122nnn/GSE122960/suppl/GSE122960_\n",
       "RAW.tar\n",
       "   title:     \n",
       "      Single-Cell Transcriptomic Analysis of Human Lung Reveals Complex Multicel\n",
       "lular Changes During Pulmonary Fibrosis II\n",
       "   type:     \n",
       "      Expression profiling by high throughput sequencing\n",
       "   web_link:     \n",
       "      https://www.atsjournals.org/doi/abs/10.1164/rccm.201712-2410OC"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "experimentData(gse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a1de951-78cc-4d4e-80ad-670831e89ab0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
